### PR TITLE
Upgrade kubebuilder version + New Alert Manager Config CRD

### PR DESCRIPTION
**Pull request by gign0766**
Upgrade kubebuilder version from v3 to v4
Add new CRD (mimiralertmanagerconfig) to handle gitops on alert manager configuration for a tenant

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
 
-COPY main.go main.go
+# Copy the go source
+COPY cmd/main.go cmd/main.go
 COPY api/ api/
-COPY controllers/ controllers/
-COPY internal/ internal/
+COPY internal/controller/ internal/controller/
 
-RUN curl -fLo mimirtool https://github.com/grafana/mimir/releases/latest/download/mimirtool-linux-amd64 && \
```
Comment (eiTi3Aishieg) [inline Dockerfile:15]: No more mimirtool in the Docker Image?

File: README.md [hunk 2]
```diff
@@ -10,73 +10,125 @@ Currently, the operator is capable of:
```
Comment (eiTi3Aishieg) [inline README.md:14]: Please add a line here in the description about the new functionality you are introducing (namely the AlertManagerConfig capabilities)

File: api/v1alpha1/mimiralertmanagerconfig_types.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (eiTi3Aishieg) [inline api/v1alpha1/mimiralertmanagerconfig_types.go]: Can we remove this?

File: api/v1alpha1/mimirrules_types.go [hunk 4]
```diff
@@ -1,10 +1,28 @@
...
+*/
+
 package v1alpha1
 
 import (
-	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (eiTi3Aishieg) [inline api/v1alpha1/mimirrules_types.go]: Can we remove this?

File: config/crd/bases/mimir.randgen.xyz_alertmanagerconfigs.yaml
Comment (eiTi3Aishieg) [inline config/crd/bases/mimir.randgen.xyz_alertmanagerconfigs.yaml:1]: I think you want this file gone, you have the same one called mimiralertmanagerconfigs
Probably forgot to remove it after renaming the CRD?

File: config/samples/kustomization.yaml [hunk 5]
```diff
@@ -1,4 +1,5 @@
-## Append samples you want in your CSV to this file as resources ##
+## Append samples of your project ##
 resources:
-- _v1alpha1_mimirrules.yaml
+  - _v1alpha1_mimirrules.yaml
+  - _v1alpha1_mimiralertmanagerconfig.yaml
```
Comment (eiTi3Aishieg) [inline config/samples/kustomization.yaml:4]: The actual file added is called _v1alpha1_alertmanagerconfig.yaml
I think you need to rename that file

File: deploy/helm/mimir-operator/crds/mimir.randgen.xyz_alertmanagerconfigs.yaml
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/crds/mimir.randgen.xyz_alertmanagerconfigs.yaml:1]: Duplicate of an already existing file but named with a "mimir" as a prefix? You probably renamed the CRD and forgot this file.

File: deploy/helm/mimir-operator/templates/rbac.yaml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
-  name: {{ include "mimir-operator.fullname" . }}
+  name: { { include "mimir-operator.fullname" . } }
```
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/templates/rbac.yaml]: might want to remove those

File: deploy/helm/mimir-operator/templates/rbac.yaml [hunk 7]
```diff
@@ -46,17 +49,16 @@ rules:
       - list
       - watch
 
-
 ---
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
 metadata:
-  name: {{ include "mimir-operator.fullname" . }}
+  name: { { include "mimir-operator.fullname" . } }
```
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/templates/rbac.yaml]: same

File: deploy/helm/mimir-operator/templates/rbac.yaml [hunk 8]
```diff
@@ -46,17 +49,16 @@ rules:
...
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
 metadata:
-  name: {{ include "mimir-operator.fullname" . }}
+  name: { { include "mimir-operator.fullname" . } }
 subjects:
   - kind: ServiceAccount
-    name:  {{ include "mimir-operator.fullname" . }}
-    namespace: {{ .Release.Namespace }}
+    name: { { include "mimir-operator.fullname" . } }
```
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/templates/rbac.yaml]: same

File: deploy/helm/mimir-operator/templates/rbac.yaml [hunk 9]
```diff
@@ -46,17 +49,16 @@ rules:
...
+    namespace: { { .Release.Namespace } }
```
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/templates/rbac.yaml]: same

File: deploy/helm/mimir-operator/templates/rbac.yaml [hunk 10]
```diff
@@ -46,17 +49,16 @@ rules:
...
 roleRef:
   kind: ClusterRole
-  name: {{ include "mimir-operator.fullname" . }}
-  apiGroup: rbac.authorization.k8s.io
\ No newline at end of file
+  name: { { include "mimir-operator.fullname" . } }
```
Comment (eiTi3Aishieg) [inline deploy/helm/mimir-operator/templates/rbac.yaml]: same

File: docs/index.md [hunk 11]
```diff
@@ -227,3 +241,64 @@ spec:
 
 The Rules installed in the Ruler by the operator will all have those labels appended to their list of labels.  
 This effectively has the same effect as overriding each individual alert to add a new label.
+
+### MimirAlertManagerConfig
+
+
+The MimirAlertManagerConfig CRD allows the remote control of the Alert Manager Config for a specific tenant in a Mimir from Kubernetes.
```
Comment (eiTi3Aishieg) [inline docs/index.md]: The MimirAlertManagerConfig CRD allows the remote control of the Alertmanager config for a specific tenant in a Mimir **instance** from Kubernetes.

Comment (eiTi3Aishieg) [reply docs/index.md]: might be clearer

File: docs/index.md [hunk 12]
```diff
@@ -227,3 +241,64 @@ spec:
...
+    your alert manager configuration in yaml
+```
+
+
+### Installing a MimirAlertManagerConfig for a Tenant
+
+In the following example, the **mimiralertmanagerconfig-sample** MimirAlertManagerConfig will be installed in the Mimir tenant under the namespace *default*.
+Then the operator will load the configuration for the tenant **tenant1** in *http://mimir.instance.com*.
+
+This example shows how to install a MimirAlertManagerConfig to configure a teams webhook into the Alert Manager:
```
Comment (eiTi3Aishieg) [inline docs/index.md]: a Teams webhook into the Alertmanager

File: internal/controller/mimiralertmanagerconfig/mimiralertmanagerconfig.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	err := mimirtool.DeleteMimirAlertManagerConfig(ctx, auth, mr.Spec.ID, mr.Spec.URL)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// configToString reads an MimirAlertManagerConfig CRD keeps only the Config Spec
+// The other fields are irrelevant to Mimir as we only need to apply the config part for the alert manager
+// func (r *MimirAlertManagerConfigReconciler) configToString(config *domain.MimirAlertManagerConfig) (string, error) {
```
Comment (eiTi3Aishieg) [inline internal/controller/mimiralertmanagerconfig/mimiralertmanagerconfig.go]: remove dead code

File: internal/controller/mimiralertmanagerconfig/mimiralertmanagerconfig_controller.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+// reconcileAMConfig ensures Mimir correctly load the alert manager config
+func (r *MimirAlertManagerConfigReconciler) reconcileAMConfig(ctx context.Context, amc *domain.MimirAlertManagerConfig) error {
+	log.FromContext(ctx).Info("Running reconciliation of the Alert Manager Config")
+
+	auth, err := utils.ExtractAuth(ctx, r.Client, amc.Spec.Auth, amc.ObjectMeta.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed to extract authentication settings: %w", err)
+	}
+
+	// config, err := r.configToString(amc)
```
Comment (eiTi3Aishieg) [inline internal/controller/mimiralertmanagerconfig/mimiralertmanagerconfig_controller.go]: dead code

**Review by eiTi3Aishieg — CHANGES_REQUESTED** [review]
Thank you for this great PR, I made a few comments for things that need changes. I'll make a test image of this to test on our internal monitoring stack and release an RC following the tests.

**Comment by gign0766** [comment]
Thank for the feedback on my PR.
I have just pushed the requested modifications.

File: Dockerfile (hunk 1 above)
Comment (gign0766) [reply Dockerfile:15]: It was a mistake and I've added it back.

**Review by eiTi3Aishieg — COMMENTED** [review]
LGTM, will merge in a separate branch to test CI before going to main

**Review by eiTi3Aishieg — APPROVED** [review]

**Pull request merged**